### ISSUE TITLE
custom ops with needs_fixed_stride_order doesn't work with auto_functionalized

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#128084]
See https://github.com/pytorch/pytorch/issues/127821#issuecomment-2147614674 for analysis.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by zou3519** [comment]
Repro:
```py
import torch
from torch._decomp import register_decomposition

lib = torch.library.Library("fsdp_test", "DEF")

lib.define("chunk_cat_(Tensor(a!) ret, Tensor[] tensors, int dim, int num_chunks) -> ()", tags=torch.Tag.needs_fixed_stride_order)

@torch.library.impl(lib, "chunk_cat_", "Meta")
def chunk_cat_(ret, tensors, dim, num_chunks):
    torch._chunk_cat(
        tensors, dim, num_chunks, out=ret
    )

@torch.library.impl(lib, "chunk_cat_", "CUDA")
def chunk_cat_(ret, tensors, dim, num_chunks):
    torch._chunk_cat(tensors, dim, num_chunks, out=ret)


def f(x, y, z):
    full_default_3: "f32[2, 524544]" = torch.ops.aten.full.default([2, 524544], 1.0, dtype = torch.float32, layout = torch.strided, device = "cuda", pin_memory = False)
    chunk_cat_default_1 = torch.ops.fsdp_test.chunk_cat_.default(full_default_3, [x, y, z], 0, 2)
    mul_out = torch.mul(full_default_3, full_default_3)
    sum_out = mul_out.sum()
    return sum_out


if __name__ == "__main__":
    device = "cuda"
    x = torch.randn([1024, 512], device=device)
    y = torch.randn([512], device=device)
    z = torch.randn([1024, 512], device=device)
    eager_out = f(x, y, z)

    compiled_aot_eager_f = torch.compile(f, backend="aot_eager", fullgraph=True)
    compiled_aot_eager_out = compiled_aot_eager_f(x, y, z)
    assert torch.allclose(eager_out, compiled_aot_eager_out), f"eager_out: {eager_out}, compiled_aot_eager_out: {compiled_aot_eager_out}"   # passes

    compiled_inductor_f = torch.compile(f, backend="inductor", fullgraph=True)
    compiled_inductor_out = compiled_inductor_f(x, y, z)
    assert torch.allclose(eager_out, compiled_inductor_out), f"eager_out: {eager_out}, compiled_inductor_out: {compiled_inductor_out}"  # fails
```

**Comment by zou3519** [comment]
Smaller repro:
```py
import torch

with torch.library._scoped_library("mylib", "DEF") as lib:

    lib.define("copy_(Tensor(a!) ret, Tensor tensors, int dim) -> ()", tags=torch.Tag.needs_fixed_stride_order)

    @torch.library.impl(lib, "copy_", "Meta")
    def _(ret, tensors, dim):
        return None

    @torch.library.impl(lib, "copy_", "CUDA")
    def _(ret, tensors, dim):
        ret.copy_(tensors)

    def f(x):
        full_default_3 = torch.full([3], 7., device="cuda")
        chunk_cat_default_1 = torch.ops.mylib.copy_.default(full_default_3, x, 0)
        mul_out = torch.mul(full_default_3, full_default_3)
        return mul_out

    device = "cuda"
    x = torch.arange(3, dtype=torch.float, device=device)
    eager_out = f(x)

    compiled_inductor_f = torch.compile(f, backend="inductor", fullgraph=True)
    compiled_inductor_out = compiled_inductor_f(x)
    assert torch.allclose(eager_out, compiled_inductor_out), f"eager_out: {eager_out}, compiled_inductor_out: {compiled_inductor_out}"  # fails
```

**Comment by zou3519** [comment]
The problem is the following. Let's say we have `needs_fixed_stride_order=True` on a `mylib.copy_` custom operator. When Inductor lowering sees the following graph:
```py
def f(x):
   y = aten.full([3], 1.)
   mylib.copy_(y, x)
   result = aten.mul(y, y)
   return result
```
it:
a) In the lowering for copy_, it coerces `y` to having the correct strides. This results it in creating a copy of `y` (call this buf0) and then calling `mylib.copy_(buf0, x)`
b) The lowering for aten.mul still uses the previous value, `y`, instead of the new buffer buf0. 

I've brainstormed two solutions.
1) Somehow get aten.mul to use the new value. This would be easy if Inductor IR had a "replace all future uses of a Node (y) with this new Node (buf0)", but I'm not sure Inductor IR was written to support that. If there's no way to support replacing users like that, then the lowering needs to add a copy_ node somewhere to write the result of buf0 back to y.
2) Change the stride order fixing to happen in the auto_functionalized lowering. What I mean by this is that Inductor actually receives the following graph before lowering:

```py
def f(x):
   y = aten.full([3], 1.)
   new_y = auto_functionalized(mylib.copy_, y, x)
   result = aten.mul(new_y, new_y)
   return result
```

If the lowering for auto_functionalized fixes the stride order of `y` by creating a clone (buf0), then calls mylib.copy_, and finally returns `buf0`, then we will be using the new value of `y` in the return. This solution would assume that all operators that have `needs_fixed_stride_order` on them are custom operators, which I think is fine (we manually handle builtin custom operators and most of those we assume have needs_fixed_stride_order=False)

Thoughts? @eellison @Chillee. My lack of understanding of Inductor IR points me toward (2).

**Comment by ezyang** [comment]
I have two intuitions here: (1) if a custom mutating op is sensitive to stride order, we probably shouldn't functionalize it, and (2) having to coerce to the right stride for a mutable arg seems pretty awful, we should try to avoid it as much as possible.

**Comment by zou3519** [comment]
> (1) if a custom mutating op is sensitive to stride order, we probably shouldn't functionalize it

AOTAutograd's "don't functionalize logic" doesn't work very well. We've run into some weird edge cases in it ([1](https://github.com/pytorch/pytorch/issues/132200), [2](https://github.com/pytorch/pytorch/pull/132243#pullrequestreview-2210730238)), and it also acts as a barrier that splits the graph and prevents fusions, which is undesirable for some of our use cases (like vllm). In general the "don't functionalize logic" is difficult to reason about.

> (2) having to coerce to the right stride for a mutable arg seems pretty awful, we should try to avoid it as much as possible.

Yup

**Comment by eellison** [comment]
I think we have two options for Tensors which are mutated and used in stride sensitive order ops.

1) At definition time of the tensor, realize the tensor and ensure that it is in the right stride. If a tensor is a view, we'd need to do this on the instantiation of its alias. It's possible that this would require a copy, such as if we had done channels last transformation.

2) Require the correct stride on the use, and if that requires a clone, copy_ back the result of the mutation afterwards. 

My preference is 1.

**Comment by malfet** [comment]
@zou3519 is there a fix for this problem? (As deadline for 2.4.1 cherry-picks are fast approaching)

**Comment by zou3519** [comment]
> At definition time of the tensor, realize it and ensure that it is in the right stride. If a tensor is a view, we'd need to do this on the instantiation of its alias. It's possible that this would require a copy, such as if we had done channels last transformation.

For the view part of this: this would need some additional infra to "invert the view". For example, if we had the following:
```py
def f(x):
    y = x[0]
    return stride_order_sensitive_op(y)
```
then we would require x[0] to have a certain stride order, but that doesn't completely constrain x's strides.

**Comment by zou3519** [comment]
Merged #133888 for 2.4.1

**Comment by atalman** [comment]
Validated release 2.4.1 with smaller repo: https://github.com/pytorch/pytorch/issues/128084#issuecomment-2265868113
